File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-5.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        if (root == null) return listOf()
        
        val result = mutableListOf<List<Int>>()
        val queue = ArrayDeque<TreeNode>(listOf(root))

        while (queue.isNotEmpty()) {
            val innerResult = mutableListOf<Int>()
            for (i in 0 until queue.size) {
                val current = queue.removeFirst()
                innerResult.add(current.`val`)
                current.left?.let { queue.add(it) }
                current.right?.let { queue.add(it) }
            }
            result.add(innerResult)
        }
        return result
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        diameter = 0

        def get_depth(node: Optional[TreeNode]) -> int:
            nonlocal diameter
            if not node:
                return 0

            left = get_depth(node.left)
            right = get_depth(node.right)

            diameter = max(diameter, left + right)

            return 1 + max(left, right)

        get_depth(root)
        return diameter